File: bindings/swift/Sources/OpenLockSwift/OpenLockTOTP.swift
import Foundation
import OpenLockFFI


/// Plaintext, unauthenticated result. Matching credentialID/timeStep only
/// correlates the response; it does not prove lock identity or physical opening.
public struct OpenLockResponse {
    public let credentialID: UInt32
    public let timeStep: UInt64
    public let errorCode: UInt32
}

/// Optional TOTP client in v3. BLE/NFC I/O and key storage are host-owned.
public enum OpenLock {
    public static func makeUnlock(secret: Data, credentialID: UInt32, unixSeconds: UInt64) throws -> Data {
        guard secret.count == 32, credentialID != 0 else { throw OpenLockError.code(3) }
        var buffer = [UInt8](repeating: 0, count: Int(OPENLOCK_REQUEST_SIZE))
        var length = 0
        let code = secret.withUnsafeBytes { key in
            openlock_make_unlock(key.bindMemory(to: UInt8.self).baseAddress, credentialID,
                                 unixSeconds, &buffer, buffer.count, &length)
        }
        guard code == 0 else { throw OpenLockError.code(code) }
        return Data(buffer[..<length])
    }

    /// Encode a code received through an authorized out-of-band path.
    public static func encodeUnlock(credentialID: UInt32, timeStep: UInt64, code: UInt32) throws -> Data {
        var buffer = [UInt8](repeating: 0, count: Int(OPENLOCK_REQUEST_SIZE))
        var length = 0
        let result = openlock_encode_unlock(credentialID, timeStep, code, &buffer, buffer.count, &length)
        guard result == 0 else { throw OpenLockError.code(result) }
        return Data(buffer[..<length])
    }

    public static func decodeResponse(_ input: Data) throws -> OpenLockResponse {
        guard input.count == Int(OPENLOCK_RESPONSE_SIZE) else { throw OpenLockError.code(3) }
        var response = openlock_response_t()
        let code = input.withUnsafeBytes { bytes in
            openlock_decode_response(bytes.bindMemory(to: UInt8.self).baseAddress, input.count, &response)
        }
        guard code == 0 else { throw OpenLockError.code(code) }
        return OpenLockResponse(credentialID: response.credential_id,
                                timeStep: response.time_step, errorCode: response.error_code)
    }
}

File: bindings/kotlin/src/main/kotlin/org/openlock/OpenLockTOTP.kt
package org.openlock


/** Plaintext, unauthenticated result; it is not proof of physical opening. */
data class OpenLockResponse(val credentialId: Long, val timeStep: Long, val errorCode: Int)

/** OpenLock TOTP client. Applications own BLE/NFC I/O and key storage. */
object OpenLock {
    init { System.loadLibrary("openlock_jni") }

    @JvmStatic private external fun makeUnlockNative(secret: ByteArray, credentialId: Long, unixSeconds: Long): ByteArray
    @JvmStatic private external fun encodeUnlockNative(credentialId: Long, timeStep: Long, code: Int): ByteArray
    @JvmStatic private external fun decodeResponseNative(input: ByteArray): LongArray

    fun makeUnlock(secret: ByteArray, credentialId: Long, unixSeconds: Long): ByteArray {
        require(secret.size == 32 && credentialId in 1..0xffff_ffffL && unixSeconds >= 0)
        return makeUnlockNative(secret, credentialId, unixSeconds)
    }

    fun encodeUnlock(credentialId: Long, timeStep: Long, code: Int): ByteArray {
        require(credentialId in 1..0xffff_ffffL && timeStep >= 0 && code in 0..99_999_999)
        return encodeUnlockNative(credentialId, timeStep, code)
    }

    fun decodeResponse(input: ByteArray): OpenLockResponse {
        require(input.size == 15)
        val fields = decodeResponseNative(input)
        return OpenLockResponse(fields[0], fields[1], fields[2].toInt())
    }
}
